atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java

```diff
@@ -324,24 +336,92 @@ private synchronized void refreshPool() {
 
     @VisibleForTesting
     void setServersInPoolTo(ImmutableSet<CassandraServer> desiredServers) {
-        ImmutableSet<CassandraServer> cachedServers = getCachedServers();
-        SetView<CassandraServer> serversToAdd = Sets.difference(desiredServers, cachedServers);
-        SetView<CassandraServer> absentServers = Sets.difference(cachedServers, desiredServers);
+        Set<CassandraServer> currentServers = getCachedServers();
+        SetView<CassandraServer> serversToAdd = Sets.difference(desiredServers, currentServers);
+        SetView<CassandraServer> absentServers = Sets.difference(currentServers, desiredServers);
 
-        serversToAdd.forEach(server -> cassandra.returnOrCreatePool(server, absentHostTracker.returnPool(server)));
         absentServers.forEach(cassandraServer -> {
             CassandraClientPoolingContainer container = cassandra.removePool(cassandraServer);
             absentHostTracker.trackAbsentCassandraServer(cassandraServer, container);
         });
 
         Set<CassandraServer> serversToShutdown = absentHostTracker.incrementAbsenceAndRemove();
 
-        if (!(serversToAdd.isEmpty() && absentServers.isEmpty())) { // if we made any changes
-            sanityCheckRingConsistency();
+        Set<CassandraServer> validatedServersToAdd =
+                validateNewHostsTopologiesAndMaybeAddToPool(getCurrentPools(), serversToAdd);
+
+        if (!(validatedServersToAdd.isEmpty() && absentServers.isEmpty())) { // if we made any changes
             cassandra.refreshTokenRangesAndGetServers();
         }
 
-        logRefreshedHosts(serversToAdd, serversToShutdown, absentServers);
+        Preconditions.checkState(
+                !getCurrentPools().isEmpty() || serversToAdd.isEmpty(),
+                "No servers were successfully added to the pool. This means we could not come to a consensus on"
+                    + " cluster topology, and the client cannot start as there are no valid hosts. This state should"
+                    + " be transient (<5 minutes), and if it is not, indicates that the user may have accidentally"
+                    + " configured AltasDB to use two separate Cassandra clusters (i.e., user-led split brain).",
+                SafeArg.of("serversToAdd", serversToAdd));
+
+        logRefreshedHosts(validatedServersToAdd, serversToShutdown, absentServers);
+    }
+
+    /**
+     * Validates new servers to add to the cassandra client container pool,
+     * by checking them with the @see {@link com.palantir.atlasdb.keyvalue.cassandra.CassandraTopologyValidator}.
```

Comment:
incorrect use of `@see` - I suggest you just remove it, the `@link` does the job.